src/win/fs-event.c

```diff
@@ -561,7 +561,28 @@ void uv__process_fs_event_req(uv_loop_t* loop, uv_req_t* req,
     }
   } else {
     err = GET_REQ_ERROR(req);
-    handle->cb(handle, NULL, 0, uv_translate_sys_error(err));
+    /*
+     * Check whether the ERROR_ACCESS_DENIED is caused by the watched directory
+     * being actually deleted (not an actual error) or a legit error.
+     */
+    if (err == ERROR_ACCESS_DENIED &&
+        handle->dirw &&
+        GetFileAttributesW(handle->dirw) == INVALID_FILE_ATTRIBUTES) {
```

Comment:
Style:
```suggestion
    if (err == ERROR_ACCESS_DENIED &&
        handle->dirw != NULL &&
        GetFileAttributesW(handle->dirw) == INVALID_FILE_ATTRIBUTES) {
```
Substance: since the check is path-based, not handle-based, isn't this basically a race condition? The directory may have been recreated, etc.

Reply:
> Substance: since the check is path-based, not handle-based, isn't this basically a race condition? The directory may have been recreated, etc.

Agreed. The problem is that checking through the handle I wasn't able to detect whether the directory was already deleted or not (`GetFileInformationByHandle()` would always succeed) so unless there's another way to detect it maybe this solution is better than what we already have?

/cc @vtjnash as he might know of a possible solution

Reply:
If you uv_stat both the current path and the handle, you could check if it is the same inode (if you still have permissions to access it)

Reply:
The problem is that checking `GetFileAttributesW()` for the current path already returns `INVALID_FILE_ATTRIBUTES` so it doesn't seem this approach would work?

Reply:
You could also try CompareObjectHandles to directly see if the new handle is the same "underlying object" as the new one, if the open succeeds

Reply:
I think I found a solution by retrieving `FileStandardInfo` from the handle. PTAL

Reply:
@bnoordhuis @vtjnash thoughts?

Reply:
Sounds interesting, but a folder could still end up not deleted (because it the user clears the flag later) and since a folder now can end up deleted without setting that flag (by requesting posix semantics for the delete, or by moving it), it would be good to test this with https://github.com/libuv/libuv/pull/4318 at least to make sure it works correctly

Reply:
@vtjnash I've updated this branch to have my changes on top of the commits in #4318. The CI looks happy. It probably doesn't cover every corner case but looks like an improvement over what we currently have. Thoughts? If you don't think it's good enough I can make a separate PR to include at least the macos fixes. Thanks.